### ISSUE TITLE
TLS support: pgagroal - PostgreSQL

### ISSUE THREAD

**Issue by jesperpedersen** [agroal/pgagroal#70]
Support TLS v1.2 and TLS v1.3.

* Extend the `server` structure with the required configuration parameters
* Extend the `connection` structure with the `SSL_SESSION` data (`i2d_SSL_SESSION()`)

**Comment by dpinchukov-ueni** [comment]
is there any hope of this getting merged into master?
i am playing around with pgagroal (1.1.0 release), and it looks like while the client->pgagroal connections can use TLS, the pgagroal->postgresql connections cannot.
support for this seems to be implemented in https://github.com/jesperpedersen/pgagroal/commit/62b14243cc6e1420028b4c4ce4e678f97c94b63a, but that branch is now rather outdated.

**Comment by jesperpedersen** [comment]
I'm working on something else at the momment, so community effort currently.

**Comment by jesperpedersen** [comment]
#404 is related

**Comment by ashu3103** [comment]
@jesperpedersen 

Please assign me this issue.

**Comment by jesperpedersen** [comment]
@ashu3103 Have fun with it !

**Comment by ashu3103** [comment]
@jesperpedersen

Currently, we have `establish_client_tls_connection` for unpooled_connections, which we are providing to the pipeline. Is it working fine? I hope there are no issues in current pipeline for ssl support.

We can however try to establish ssl connections in `_get_connection` and store that ssl context in `config->connections[slot]` as suggested above which can be further utilised in our pipelines.

Any views on this?

Thanks

**Comment by ashu3103** [comment]
@jesperpedersen 

What is the issue with the current implementation, the worker is establishing tls for itself and supplies the ssl context to pipeline for ssl support in further connections.

Just for the part where `use_pooled_connection` (reusing a prefilled connection) are used there is no ssl.

**Comment by jesperpedersen** [comment]
@ashu3103 Yeah, we can connect to a TLS protected PostgreSQL instance, but the connections can't be cached since we can't currently serialize the SSLContext to be part of the pool.

You can investigate newer versions of OpenSSL, or look at alternative implmentations like 

* https://github.com/eduardsui/tlse
* https://github.com/Mbed-TLS/mbedtls

Create a test setup with self-signed certificates, create a connection, and return it to the pool. You will see it isn't cached.

This ties into requiring TLS certificates for connections - it is the easier part to start with

**Comment by ashu3103** [comment]
> @ashu3103 Yeah, we can connect to a TLS protected PostgreSQL instance, but the connections can't be cached since we can't currently serialize the SSLContext to be part of the pool.
> 

But why are we not able to make SSLContext a part of the pool?

> You can investigate newer versions of OpenSSL, or look at alternative implmentations like
> 
> * https://github.com/eduardsui/tlse
> * https://github.com/Mbed-TLS/mbedtls
> 

Sure

> Create a test setup with self-signed certificates, create a connection, and return it to the pool. You will see it isn't cached.
> 
> This ties into requiring TLS certificates for connections - it is the easier part to start with

Okay!

**Comment by ashu3103** [comment]
> > @ashu3103 Yeah, we can connect to a TLS protected PostgreSQL instance, but the connections can't be cached since we can't currently serialize the SSLContext to be part of the pool.
> 
> But why are we not able to make SSLContext a part of the pool? Something like `config->connections[ssl] = (SSL*)ssl_context`

In `_prefill_auth` we can do a ssl handshake after `_get_connection` and save in pool

> 
> > You can investigate newer versions of OpenSSL, or look at alternative implmentations like
> > 
> > * https://github.com/eduardsui/tlse
> > * https://github.com/Mbed-TLS/mbedtls
> 
> Sure, but don't you think openssl has better community support and therefore rapid development
> 
> > Create a test setup with self-signed certificates, create a connection, and return it to the pool. You will see it isn't cached.
> > This ties into requiring TLS certificates for connections - it is the easier part to start with
> 
> Okay!

**Comment by jesperpedersen** [comment]
We have to reserve the slot in the pool - even for a SSL based connection
